nntrainer/layers/layer_impl.h

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   layer_impl.h
+ * @date   21 June 2021
+ * @brief  This is base Optimizer implementation class
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#ifndef __LAYER_IMPL_H__
+#define __LAYER_IMPL_H__
+#ifdef __cplusplus
+
+#include <layer_devel.h>
+
+#include <memory>
+#include <tuple>
+
+namespace nntrainer {
+
+class InitContext;
+class RunContext;
+class Exporter;
+
+enum class ExportMethods;
+/**
+ * @class   An abstract class to ease developing a layer
+ * @brief   An abstract class for all layers
+ *
+ */
+class LayerImpl : public Layer {
+
+public:
+  /**
+   * @brief     Constructor of Layer Class
+   */
+  LayerImpl();
+
+  /**
+   * @brief     Destructor of Layer Class
+   */
+  virtual ~LayerImpl() = default;
+
+  /**
+   * @brief     finalize the layer
+   * @throw     nntrainer::not_supported if try to initialize twice
+   * @copydoc   Layer::fianlize(InitContext &context)
+   */
+  virtual void finalize(InitContext &context) override;
+
+  /**
+   * @copydoc Layer::setProperty(const std::vector<std::string> &values)
+   */
+  virtual void setProperty(const std::vector<std::string> &values) override;
+
+  /**
+   * @copydoc Layer::exportTo(Exporter &exporter, const ExportMethods &methods)
+   */
+  virtual void exportTo(Exporter &exporter,
+                        const ExportMethods &method) const override;
+
+private:
+  bool finalized;                                 /**< check if finalized */
+  std::unique_ptr<std::tuple<>> layer_impl_props; /**< layer_impl_props */
+};
+
+/**
+ * @brief General Layer Factory function to register Layer
+ *
+ * @param props property representation
+ * @return std::unique_ptr<ml::train::Layer> created object
+ */
+template <typename T,
```

Comment:
why move to LayerImpl? this will be called by model/graph/etc and they shouldnt have to include layer_impl.

Reply:
I think the specification is the signature `std::unique_ptr<Layer>(const std::vector<std::string &props>)`. 

If we are going to make this implementation as a specification of an api, works for me.  But I guess we can delay this until we need it (maybe soon 😛 )

Reply:
isn't what the API calls? its left up to you where you keep this but no other header should import layer_impl.h except the ones which derive from LayerImpl (if they do, then a lot other headers will need importing as well). Others should import layer_devel.h

Reply:
Yes. I'll keep in mind, if this is not possible this must go to layer_devel.h.

Reply:
As discussed offline, this is reverted back